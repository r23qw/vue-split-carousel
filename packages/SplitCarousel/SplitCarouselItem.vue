<script lang="ts">
import {
  computed,
  defineComponent,
  getCurrentInstance,
  inject,
  onMounted,
  onUnmounted,
  ref,
} from 'vue-demi'
import type { ComponentStyle, InjectCarouselScope } from './carousel'

export default defineComponent({
  name: 'SplitCarouselItem',
  setup() {
    let instance = getCurrentInstance()!
    let uid = Math.random()
    const injectCarouselScope: InjectCarouselScope = inject(
      'injectCarouselScope',
    ) as InjectCarouselScope
    const isMounted = ref(false)

    if (!instance || !injectCarouselScope)
      throw new Error('SplitCarouselItem Must include by SplitCarousel')

    const { layout, stag, reset, addItem, removeItem } = injectCarouselScope
    // register/unregister
    onMounted(() => {
      addItem({ uid })
      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          isMounted.value = true
        })
      })
    })
    onUnmounted(() => {
      removeItem(uid)
    })

    // itemStyle
    const isPrepend = computed(() => uid === stag.value.prependUid)
    const isAppend = computed(() => uid === stag.value.appendUid)
    const stagIndex = computed(() => {
      return stag.value.stagUids.findIndex(id => id === uid)
    })

    const inStag = computed(() => stagIndex.value !== -1)
    const noAnimate = computed(() => {
      if (!isMounted.value)
        return true

      if (reset.resetting)
        return true

      return false
    })

    const itemStyle = computed(() => {
      const {
        itemWidth,
        isStatic,
        itemBlockWidth,
        appendPostion,
        prependPosition,
      } = layout.value

      let style: ComponentStyle = {
        width: `${itemWidth}px`,
        transitionDuration: `${noAnimate.value ? 0 : injectCarouselScope.speed.value
          }ms`,
        transitionTimingFunction: injectCarouselScope.timingFunction.value,
      }

      if (isStatic) {
        style.display = 'block'
        return style
      }

      if (inStag.value) {
        style = {
          ...style,
          display: 'block',
          transform: `translateX(${stagIndex.value * itemBlockWidth}px)`,
        }
      }
      if (isPrepend.value) {
        const needReset = reset.resetting && reset.action === 'next'
        style = {
          ...style,
          display: 'block',
          transform: `translateX(${needReset ? appendPostion : prependPosition
            }px)`,
        }
      }
      if (isAppend.value) {
        const needReset = reset.resetting && reset.action === 'prev'
        style = {
          ...style,
          display: 'block',
          transform: `translateX(${needReset ? prependPosition : appendPostion
            }px)`,
        }
      }

      return style
    })

    return {
      itemStyle,
      layout,
    }
  },
})
</script>

<template>
  <div class="split-carousel__item" :class="{ 'split-carousel__item--transition': !layout.isStatic }" :style="itemStyle">
    <slot />
  </div>
</template>

<style scoped lang="scss">
.split-carousel__item {
  display: none;
  transition-property: transform;
  will-change: transfrom;

  &--transition {
    position: absolute;
    left: 0;
    top: 0;
    height: 100%;
  }

  &--stag {
    display: block;
  }
}
</style>
