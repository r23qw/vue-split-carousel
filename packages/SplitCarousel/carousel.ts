import type { ComputedRef, Ref } from 'vue-demi'


export interface CarouselItem {
  uid: number
}

export interface CarouselStag {
  prependUid: number
  stagUids: number[]
  appendUid: number
}

export interface CarouselReset {
  resetting: boolean
  action: 'prev' | 'next'
}

export interface CarouselLayout {
  isStatic: boolean
  viewportWidth: number
  itemWidth: number
  gapWidth: number
  itemBlockWidth: number
  prependPosition: number
  appendPostion: number
}

export interface InjectCarouselScope {
  stag: ComputedRef<CarouselStag>
  reset: CarouselReset
  layout: ComputedRef<CarouselLayout>
  removeItem: (id: number) => void
  addItem: (item: CarouselItem) => void
  speed: Ref<number>
  timingFunction: Ref<string>
}

export interface ComponentStyle {
  [key: string]: string
}
