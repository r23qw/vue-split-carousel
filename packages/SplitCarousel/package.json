{
  "name": "split-carousel",
  "version": "3.0.0",
  "description": "",
  "main": "index.ts",
  "scripts": {},
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "vue": "^3.2.47"
  }
}
