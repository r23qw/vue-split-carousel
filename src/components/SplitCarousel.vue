<template>
  <div
    class="split-carousel"
    :style="{
      height: `${typeof height === 'string' ? height : `${height}px`}`,
    }"
  >
    <div class="split-carousel__left" @click="prev">
      <div v-show="isLeftArrowVisiable">
        <slot v-if="hasLeftSlot" name="left"></slot>
        <button class="split-carousel__left-button" v-else>
          <div class="arrow left"></div>
        </button>
      </div>
    </div>
    <div
      ref="viewportDOMRef"
      class="split-carousel__viewport"
      @mouseenter="enter"
      @mouseleave="leave"
      :class="{ 'split-carousel__viewport--static': layout.isStatic }"
    >
      <slot></slot>
    </div>
    <div class="split-carousel__right" @click="next">
      <div v-show="isRightArrowVisiable">
        <slot v-if="hasRightSlot" name="right"></slot>
        <button class="split-carousel__right-button" v-else>
          <div class="arrow right"></div>
        </button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import {
  computed,
  ComputedRef,
  defineComponent,
  nextTick,
  onMounted,
  onUnmounted,
  provide,
  reactive,
  ref,
  toRefs,
  watch,
} from "vue";
import type {
  CarouselItem,
  CarouselLayout,
  CarouselReset,
  ComponentUid,
  InjectCarouselScope,
} from "./carousel";

export default defineComponent({
  name: "SplitCarousel",
  props: {
    autoplay: {
      type: Boolean,
      default: true,
    },
    speed: {
      type: Number,
      default: 500,
    },
    interval: {
      type: Number,
      default: 3000,
    },
    loop: {
      type: Boolean,
      default: true,
    },
    displayAmount: {
      type: Number,
      default: 4,
    },
    height: {
      type: [Number, String],
      default: 80,
    },
    itemWidth: {
      type: Number,
      default: 120,
    },
    pauseOnHover: {
      type: Boolean,
      default: true,
    },
    timingFunction: {
      type: String,
      default: "ease",
    },
    arrowVisible: {
      type: String,
      validator: (i: string) => ["default", "always"].includes(i),
      default: "default",
    },
  },
  setup(props, context) {
    const viewportDOMRef = ref<null | HTMLElement>(null);

    const items = ref<CarouselItem[]>([]);
    const addItem = (item: CarouselItem) => items.value.push(item);
    const removeItem = (uid: ComponentUid) => {
      const index = items.value.findIndex((item) => item.uid === uid);
      if (index !== -1) {
        items.value.splice(index, 1);
      }
    };
    // layout
    const viewportWidth = ref(0);
    const layout: ComputedRef<CarouselLayout> = computed(() => {
      const gapWidth =
        (viewportWidth.value - props.itemWidth * props.displayAmount) /
        (props.displayAmount - 1);
      const itemBlockWidth = gapWidth + props.itemWidth;
      return {
        isStatic: items.value.length <= props.displayAmount,
        gapWidth,
        viewportWidth: viewportWidth.value,
        itemWidth: props.itemWidth,
        itemBlockWidth,
        prependPosition: -1 * itemBlockWidth,
        appendPostion: viewportWidth.value + gapWidth,
      };
    });

    // index
    const activeIndex = ref(0);
    const isLastIndex = computed(
      () => activeIndex.value + props.displayAmount === items.value.length
    );
    const isFirstIndex = computed(() => activeIndex.value === 0);
    const isNeedReset = computed(
      () => props.displayAmount + 2 >= items.value.length
    );
    const reset = reactive<CarouselReset>({
      resetting: false,
      action: "next",
    });

    const stag = computed(() => {
      const index = activeIndex.value;
      const list = items.value;
      const endIndex = list.length - 1;

      const stagUids = Array.from({ length: props.displayAmount }).map(
        (v, offset) => list[(index + offset) % list.length]?.uid
      );

      const result = {
        prependUid: list[isFirstIndex.value ? endIndex : index - 1]?.uid,
        stagUids,
        appendUid: list[(index + props.displayAmount) % list.length]?.uid,
      };
      if (result.prependUid === result.appendUid) {
        const invalidUid = -1;
        if (reset.action === "prev") {
          result.prependUid = invalidUid;
        }
        if (reset.action === "next") {
          result.appendUid = invalidUid;
        }
      }
      return result;
    });
    const refsProps = toRefs(props);
    provide<InjectCarouselScope>("injectCarouselScope", {
      speed: refsProps.speed,
      timingFunction: refsProps.timingFunction,
      stag,
      reset,
      addItem,
      removeItem,
      layout,
    });

    //arrow
    const isLeftArrowVisiable = computed(() => {
      if (props.arrowVisible === "default") {
        if (layout.value.isStatic) {
          return false;
        }
        if (isFirstIndex.value && !props.loop) {
          return false;
        }
      }
      return true;
    });
    const isRightArrowVisiable = computed(() => {
      if (props.arrowVisible === "default") {
        if (layout.value.isStatic) {
          return false;
        }
        if (isLastIndex.value && !props.loop) {
          return false;
        }
      }
      return true;
    });

    // play method
    const slide = (action: CarouselReset["action"]) => {
      if (layout.value.isStatic) return;
      if (!props.loop) {
        if (action === "next" && isLastIndex.value) return;
        if (action === "prev" && isFirstIndex.value) return;
      }

      const setIndex = () => {
        if (action === "next") {
          activeIndex.value = (activeIndex.value + 1) % items.value.length;
        }
        if (action === "prev") {
          activeIndex.value = isFirstIndex.value
            ? items.value.length - 1
            : activeIndex.value - 1;
        }
      };

      if (!isNeedReset.value) {
        setIndex();
        return;
      }

      reset.resetting = true;
      reset.action = action;
      requestAnimationFrame(() => {
        // this frame setting edge item position
        requestAnimationFrame(() => {
          reset.resetting = false;
          setIndex();
        });
      });
    };
    let timer: ReturnType<typeof setTimeout>;
    const next = () => {
      clearTimeout(timer);
      slide("next");
      if (props.autoplay) {
        timer = setTimeout(() => next(), props.interval);
      }
    };
    const prev = () => {
      clearTimeout(timer);
      slide("prev");
      if (props.autoplay) {
        timer = setTimeout(() => prev(), props.interval);
      }
    };
    const enter = () => {
      if (props.autoplay && props.pauseOnHover) {
        clearTimeout(timer);
      }
    };
    const leave = () => {
      if (props.autoplay && props.pauseOnHover) {
        clearTimeout(timer);
        timer = setTimeout(() => next(), props.interval);
      }
    };

    // autoplay
    watch(
      () => props.autoplay,
      (autoplay) => {
        if (autoplay) {
          timer = setTimeout(() => next(), props.interval);
        } else {
          clearTimeout(timer);
        }
      }
    );
    const handlePageVisiblityChange = () => {
      if (document.visibilityState === "hidden") {
        clearTimeout(timer);
      }
      if (document.visibilityState === "visible" && props.autoplay) {
        timer = setTimeout(() => next(), props.interval);
      }
    };

    // init layout,handle page visibility
    onMounted(() => {
      nextTick(() => {
        if (viewportDOMRef.value !== null) {
          viewportWidth.value = viewportDOMRef.value.clientWidth;
        }

        if (props.autoplay) {
          timer = setTimeout(() => next(), props.interval);
        }

        document.addEventListener(
          "visibilitychange",
          handlePageVisiblityChange
        );
      });
    });
    onUnmounted(() => {
      clearTimeout(timer);

      document.removeEventListener(
        "visibilitychange",
        handlePageVisiblityChange
      );
    });

    return {
      prev,
      next,
      enter,
      leave,
      layout,
      viewportDOMRef,
      hasLeftSlot: context.slots.left !== undefined,
      hasRightSlot: context.slots.right !== undefined,
      isLeftArrowVisiable,
      isRightArrowVisiable,
    };
  },
});
</script>

<style scoped lang="scss">
.split-carousel {
  display: flex;
  width: 100%;
  align-items: center;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  &__viewport {
    flex-grow: 1;
    overflow: hidden;
    align-self: stretch;
    position: relative;
    &--static {
      display: flex;
      align-items: center;
      justify-content: space-between;
      ::v-deep .split-carousel__item {
        align-self: stretch;
      }
    }
  }
  &__left,
  &__right {
    align-self: center;
    padding: 5px;
  }
  &__left-button,
  &__right-button {
    height: 30px;
    width: 30px;
    text-align: center;
    line-height: 30px;
    border: 1px solid #dcdfe6;
    border-radius: 50%;
    cursor: pointer;
    background-color: #fff;
    transition: border-color, background-color 0.15s;
    display: flex;
    align-items: center;
    justify-content: center;
    &:hover {
      border-color: #c6e2ff;
      background-color: #ecf5ff;
      .arrow {
        border-top-color: #c6e2ff;
        border-left-color: #c6e2ff;
      }
    }
  }
}
.arrow {
  height: 6px;
  width: 6px;
  border: 2px solid;
  transition: border-color 0.15s;
  border-top-color: #dcdfe6;
  border-left-color: #dcdfe6;
  border-right-color: transparent;
  border-bottom-color: transparent;
  &.left {
    transform: translateX(2px) rotate(-45deg);
  }
  &.right {
    transform: translateX(-2px) rotate(135deg);
  }
}
</style>
