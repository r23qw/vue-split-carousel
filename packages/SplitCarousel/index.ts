import type { App } from 'vue-demi';
import SplitCarousel from './SplitCarousel.vue';
import SplitCarouselItem from './SplitCarouselItem.vue';

interface SplitCarouselOption {
  prefix: string;
}

const install = (app: App, option?: SplitCarouselOption): void => {
  const prefix = option?.prefix || '';
  app.component(`${prefix}SplitCarousel`, SplitCarousel as any);
  app.component(`${prefix}SplitCarouselItem`, SplitCarouselItem as any);
};

export { SplitCarousel, SplitCarouselItem, install };
export default install;

declare module 'vue' {
  export interface GlobalComponents {
    SplitCarousel: typeof SplitCarousel;
    SplitCarouselItem: typeof SplitCarouselItem;
  }
}
