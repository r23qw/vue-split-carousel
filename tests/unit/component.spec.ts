import { mount } from "@vue/test-utils";
import { defineComponent } from "vue";
import SplitCarousel from "@/components/SplitCarousel.vue";
import SplitCarouselItem from "@/components/SplitCarouselItem.vue";

const sleep = (interval: number) =>
  new Promise((resolve) => {
    setTimeout(resolve, interval);
  });

describe("Split Carousel Component", () => {
  const CarouselWrapper = defineComponent({
    props: {
      itemAmount: {
        type: Number,
        default: 6,
      },
      option: {
        type: Object,
        default: () => ({
          displayAmount: 4,
          autoplay: true,
          speed: 500,
          interval: 3000,
          loop: true,
          height: 80,
          itemWidth: 120,
          pauseOnHover: true,
          timingFunction: "ease",
          arrowVisible: "default",
        }),
      },
    },
    components: { SplitCarousel, SplitCarouselItem },
    template: `
      <split-carousel v-bind="option">
        <split-carousel-item v-for="item in itemAmount" :key="item">
          {{item}}
        </split-carousel-item>
      </split-carousel>
    `,
  });
  it("Match snapshot", () => {
    const wrapper = mount(CarouselWrapper);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("Static carousel", () => {
    const wrapper = mount(CarouselWrapper, { props: { itemAmount: 3 } });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("Autoplay", async () => {
    const wrapper = mount(CarouselWrapper);
    await sleep(4000);
    const first = wrapper.findAllComponents(SplitCarouselItem).filter((i) => {
      return (i.vm as any).itemStyle.transform === "translateX(0px)";
    });
    expect(first?.[0].text()).toBe("2");
  });
  it("Not autoplay", async () => {
    const wrapper = mount(CarouselWrapper, {
      props: { option: { autoplay: false } },
    });
    await sleep(4000);
    const first = wrapper.findAllComponents(SplitCarouselItem).filter((i) => {
      return (i.vm as any).itemStyle.transform === "translateX(0px)";
    });
    expect(first?.[0].text()).toBe("1");
  });
});
