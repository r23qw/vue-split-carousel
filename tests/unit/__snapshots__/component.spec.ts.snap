// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Split Carousel Component Match snapshot 1`] = `
<div class="split-carousel" style="height: 80px;">
  <div class="split-carousel__left">
    <div style="display: none;"><button class="split-carousel__left-button">
        <div class="arrow left"></div>
      </button></div>
  </div>
  <div class="split-carousel__viewport split-carousel__viewport--static">
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">1</div>
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">2</div>
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">3</div>
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">4</div>
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">5</div>
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">6</div>
  </div>
  <div class="split-carousel__right">
    <div style="display: none;"><button class="split-carousel__right-button">
        <div class="arrow right"></div>
      </button></div>
  </div>
</div>
`;

exports[`Split Carousel Component Static carousel 1`] = `
<div class="split-carousel" style="height: 80px;">
  <div class="split-carousel__left">
    <div style="display: none;"><button class="split-carousel__left-button">
        <div class="arrow left"></div>
      </button></div>
  </div>
  <div class="split-carousel__viewport split-carousel__viewport--static">
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">1</div>
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">2</div>
    <div class="split-carousel__item" style="width: 120px; transition-duration: 0ms; transition-timing-function: ease; display: block;">3</div>
  </div>
  <div class="split-carousel__right">
    <div style="display: none;"><button class="split-carousel__right-button">
        <div class="arrow right"></div>
      </button></div>
  </div>
</div>
`;
